minima/database.nim

```diff
@@ -20,38 +20,55 @@ type
         KeyNotFound             = "minima: key not found"
         PersistenceFailed       = "minima: persistence failed"
 
-proc log(db: Database, key: seq[byte], value: seq[byte]) =
-    let write = concat(
-        @(uint32(len(key)).toBytes),
-        @(uint32(len(value)).toBytes),
-        key,
-        value
-    )
-
-    # @TODO CATCH EXCEPTIONS
-    discard db.log.writeBytes(write, 0, len(write))
-    db.log.flushFile()
+proc toAESKey*(str: string): array[aes256.sizeKey, byte] =
+    ## Concerts a string to an AES Key for opening an encrypted database.
+    ## 
+    ## **Example::**
+    ## .. code-block::
+    ##   let password = "foo".toAESKey
+    var pass = str
+    var key: array[32, byte]
+    copyMem(addr key[0], addr pass[0], len(pass))
 
-proc readInt(file: File): int =
-    var arr: array[4, byte]
-    discard file.readBytes(arr, 0, 4)
+    return key
 
-    return int(uint32.fromBytes(arr))
+proc open*(dir: string, key: array[aes256.sizeKey, byte]): Result[Database, DatabaseError] =
+    ## Opens an encrypted database at the specified path.
+    ## 
+    ## **Example:**
+    ##
+    ## .. code-block::
+    ##   let result = open("/tmp/minima", "password".toAESKey)
+    ##   if not result.isOk:
+    ##     return
+    if not os.existsDir(dir):
+        try:
+            os.createDir(dir)
+        except:
+            return err(DirectoryCreationFailed)
 
-proc recover(db: Database) =
-    while db.log.getFilePos() <= db.log.getFileSize() - 1:
-        var keyLen = readInt(db.log)
-        var valLen = readInt(db.log)
+    var path = dir & "/minima.db"
+    var mode = fmReadWrite
+    if fileExists(path):
+        mode = fmReadWriteExisting
 
-        # @TODO CATCH EXCEPTIONS
+    var f: File
+    try:
+        f = open(path, mode)
+    except:
+        return err(TreeFileCreationFailed)
 
-        var key = newSeq[byte](keyLen)
-        discard db.log.readBytes(key, 0, keyLen)
+    var db = Database(
+        dir: dir,
+        log: EncryptedLog.init(f, key),
+        tree: initBTree[string, seq[byte]]()
```

Comment:
In our own libraries we follow a more "modern" convention for `init` functions:

Each init function should accept a single type argument specifying the constructed type:

```nim
proc init*[A, B](T: type BTree[A, B]): T =
  ...
```

More examples here:
https://github.com/status-im/nim-stew/blob/master/stew/shims/tables.nim#L3


This allows the creation of some generic helpers related to initialization. For example, if you had a field or a variable that you want to initialize, you can rely on API that infers the type of the initialized location:

```nim
proc foo(): SomeTreeHolder =
  ...
  init result.tree
```

In the future, the language will gain more support for a method of type inference called [target typing](https://anthonygiretti.com/2020/06/21/introducing-c-9-improved-target-typing/). It will bring additional simplifications when this convention is used.